include/lsst/afw/cameraGeom/Amplifier.h

```diff
@@ -101,188 +95,268 @@ class Amplifier {
+
+protected:
+
+    struct Fields {
+        std::string name;
+        lsst::geom::Box2I bbox;
+        double gain = 0.0;
+        double readNoise = 0.0;
+        double saturation = 0.0;
+        double suspectLevel = 0.0;
+        ReadoutCorner readoutCorner = ReadoutCorner::LL;
+        ndarray::Array<double const, 1, 1> linearityCoeffs;
+        std::string linearityType;
+        double linearityThreshold;
+        double linearityMaximum;
+        std::string linearityUnits;
+        lsst::geom::Box2I rawBBox;
+        lsst::geom::Box2I rawDataBBox;
+        bool rawFlipX = false;
+        bool rawFlipY = false;
+        lsst::geom::Extent2I rawXYOffset;
+        lsst::geom::Box2I rawHorizontalOverscanBBox;
+        lsst::geom::Box2I rawVerticalOverscanBBox;
+        lsst::geom::Box2I rawPrescanBBox;
+    };
+
+    Amplifier() = default;
+
+    Amplifier(Amplifier const &) = default;
+    Amplifier(Amplifier &&) = default;
+    Amplifier & operator=(Amplifier const &) = default;
+    Amplifier & operator=(Amplifier &&) = default;
+
+    virtual Fields const & getFields() const = 0;
+
+};
+
+/**
+ *  A mutable Amplifier subclass class that can be used to incrementally
+ *  construct or modify Amplifiers.
+ */
+class Amplifier::Builder final : public Amplifier {
+public:
+
+    /**
+     * Construct a new Builder object from the fields in the given
+     * record.
+     *
+     * @param[in] record   Record to copy fields from.
+     */
+    static Builder fromRecord(table::BaseRecord const & record);
+
+    /// Construct a Builder with default values for all fields.
+    Builder() = default;
+
+    /// Standard copy constructor.
+    Builder(Builder const &) = default;
+
+    /// Standard move constructor.
+    Builder(Builder &&) = default;
+
+    /// Construct a Builder with values initialized from the given Amplifier.
+    Builder(Amplifier const & other);
+
+    /// Standard copy assignment.
+    Builder & operator=(Builder const &) = default;
+
+    /// Standard move assignment.
+    Builder & operator=(Builder &&) = default;
+
+    /// Set the Builder's fields to those of the given Amplifier.
+    Builder & operator=(Amplifier const & other);
+
+    ~Builder() noexcept override = default;
+
+    /**
+     *  Construct an immutable Amplifier with the same values as the Builder.
+     *
+     *  The derived type of the return instance is unspecified, and should be
+     *  considered an implementation detail.
+     */
+    std::shared_ptr<Amplifier const> finish() const;
+
+    /// @copydoc Amplifier::getName
+    void setName(std::string const &name) { _fields.name = name; }
+
+    /// @copydoc Amplifier::getBBox
+    void setBBox(lsst::geom::Box2I const &bbox) { _fields.bbox = bbox; }
+
+    /// @copydoc Amplifier::getGain
+    void setGain(double gain) { _fields.gain = gain; }
+
+    /// @copydoc Amplifier::getReadNoise
+    void setReadNoise(double readNoise) { _fields.readNoise = readNoise; }
```

Comment:
We can definitely wrap that kind of code in pybind11, since both C++ and Python convention require assignment operators (in particular, compound assignments) to do this, and we do have some wrappers like that in `afw` (`afw.math`, I think?). So any difficulties would be social rather than technical.

Reply:
My hope is that this kind of amplifier construction never occurs, and that the builders are hidden behind user friendly functions.  My expected user interaction cases are like those in `obs_cfht` and `obs_lsst`, where due to peculiarities of a camera, a small number of fields (readNoise and BBox, for these examples) need to be reassigned based on some specific logic.